File: CS/E3422/DialogService.cs
using System;
using System.IO;
using System.Windows;
using System.Windows.Browser;
using System.Windows.Controls;
using DevExpress.DocumentServices.ServiceModel.Native;
using DevExpress.Xpf.Printing;
using DevExpress.Xpf.Printing.Native;

namespace E3422 {
    public class DialogService : IDialogService {
        public void ShowPreview(string serviceUri, string reportName) {
            ReportPreviewModel model = new ReportPreviewModel(serviceUri);
            model.ReportName = reportName;
            DocumentPreviewWindow preview = new DocumentPreviewWindow() { Model = model };
            model.CreateDocument();
            preview.ShowDialog();
        }

        public Stream ShowSaveFileDialog(string filter) {
            SaveFileDialog dialog = new SaveFileDialog();
            dialog.Filter = filter;
       
            if(dialog.ShowDialog() != true)
                return null;

            return dialog.OpenFile();
        }

        public void ShowMessage(string caption, string message) {
            MessageBox.Show(message, caption ?? string.Empty, MessageBoxButton.OK);
        }

        public void OpenBrowserWindow(Uri uri) {
            HtmlPage.Window.Navigate(uri, "_blank", "toolbar=0,menubar=0,resizable=1,scrollbars=1");
        }

        public void AsyncRequestPrintingConfirmation(Action<bool> continuePrinting) {
            LoadingPrintDataWindow window = new LoadingPrintDataWindow();

            LoadingPrintDataViewModel model = new LoadingPrintDataViewModel();
            model.SetStatus(PrintingStatus.Generated);
            model.Continue += (s, a) => {
                window.Hide();
                continuePrinting(true);
            };
            model.Cancel += (s, a) => {
                window.Hide();
                continuePrinting(false);
            };

            window.ViewModel = model;
            window.Show();
        }
    }
}

File: VB/E3422/DialogService.vb
Imports System
Imports System.IO
Imports System.Windows
Imports System.Windows.Browser
Imports System.Windows.Controls
Imports DevExpress.DocumentServices.ServiceModel.Native
Imports DevExpress.Xpf.Printing
Imports DevExpress.Xpf.Printing.Native

Namespace E3422
    Public Class DialogService
        Implements IDialogService

        Public Sub ShowPreview(ByVal serviceUri As String, ByVal reportName As String) Implements IDialogService.ShowPreview
            Dim model As New ReportServicePreviewModel(serviceUri)
            model.ReportName = reportName
            Dim preview As New DocumentPreviewWindow() With {.Model = model}
            model.CreateDocument()
            preview.ShowDialog()
        End Sub

        Public Function ShowSaveFileDialog(ByVal filter As String) As Stream Implements IDialogService.ShowSaveFileDialog
            Dim dialog As New SaveFileDialog()
            dialog.Filter = filter

            If dialog.ShowDialog() <> True Then
                Return Nothing
            End If

            Return dialog.OpenFile()
        End Function

        Public Sub ShowMessage(ByVal caption As String, ByVal message As String) Implements IDialogService.ShowMessage
            MessageBox.Show(message, If(caption, String.Empty), MessageBoxButton.OK)
        End Sub

        Public Sub OpenBrowserWindow(ByVal uri As Uri) Implements IDialogService.OpenBrowserWindow
            HtmlPage.Window.Navigate(uri, "_blank", "toolbar=0,menubar=0,resizable=1,scrollbars=1")
        End Sub

        Public Sub AsyncRequestPrintingConfirmation(ByVal continuePrinting As Action(Of Boolean)) Implements IDialogService.AsyncRequestPrintingConfirmation
            Dim window As New LoadingPrintDataWindow()

            Dim model As New LoadingPrintDataViewModel()
            model.SetStatus(PrintingStatus.Generated)
            AddHandler model.Continue, Sub(s, a)
                window.Hide()
                continuePrinting(True)
            End Sub
            AddHandler model.Cancel, Sub(s, a)
                window.Hide()
                continuePrinting(False)
            End Sub

            window.ViewModel = model
            window.Show()
        End Sub
    End Class
End Namespace
